pkg/protoassert/assert.go

```diff
@@ -1,130 +1,194 @@
 	t.Helper()
 	expectedKeys := maps.Keys(expected)
 	actualKeys := maps.Keys(actual)
-	areEqual := !assert.ElementsMatch(t, expectedKeys, actualKeys)
-	for expectedKey, expectedValue := range expected {
-		a := actual[expectedKey]
-		areEqual = SlicesEqual(t, expectedValue, a, expectedKey) && areEqual
+	if !assert.ElementsMatch(t, expectedKeys, actualKeys, msgAndArgs...) {
+		return false
 	}
-	if !areEqual {
-		t.Log(msgAndArgs...)
+	for key, expectedV := range expected {
+		actualV := actual[key]
+		if !SlicesEqual(t, expectedV, actualV, msgAndArgs...) {
+			return false
+		}
 	}
-	return areEqual
+	return true
 }
 
-func MapEqual[K comparable, V any, T message[V]](t testing.TB, expected, actual map[K]T, msgAndArgs ...interface{}) bool {
+// MapEqual determines if the expected and actual maps are equal.
+func MapEqual[M ~map[K]T, K comparable, T message[V], V any](t testing.TB, expected, actual M, msgAndArgs ...any) bool {
 	t.Helper()
 	expectedKeys := maps.Keys(expected)
 	actualKeys := maps.Keys(actual)
-	areEqual := !assert.ElementsMatch(t, expectedKeys, actualKeys)
-	for expectedKey, expectedValue := range expected {
-		actualValue := actual[expectedKey]
-		areEqual = Equal(t, expectedValue, actualValue, expectedKey) && areEqual
+	if !assert.ElementsMatch(t, expectedKeys, actualKeys, msgAndArgs...) {
+		return false
 	}
-	if !areEqual {
-		t.Log(msgAndArgs...)
+	for key, expectedV := range expected {
+		actualV := actual[key]
+		if !Equal(t, expectedV, actualV, msgAndArgs...) {
+			return false
+		}
 	}
-	return areEqual
+	return true
 }
 
-func toJson(m proto.Message) (string, error) {
-	if m == nil {
-		return "", nil
+// diffSlices is based on [assert.diffLists]. The doc for it is copied below:
+//
+// diffLists diffs two arrays/slices and returns slices of elements that are only in A and only in B.
+// If some element is present multiple times, each instance is counted separately (e.g. if something is 2x in A and
+// 5x in B, it will be 0x in extraA and 3x in extraB). The order of items in both lists is ignored.
+func diffSlices[V any, T message[V]](a, b []T) (extraA, extraB []T) {
+	aLen, bLen := len(a), len(b)
+
+	visited := make([]bool, bLen)
+	for i := 0; i < aLen; i++ {
+		element := a[i]
+		found := false
+		for j := 0; j < bLen; j++ {
+			if visited[j] {
+				continue
+			}
+			if b[j].EqualVT(element) {
+				visited[j] = true
+				found = true
+				break
+			}
+		}
+		if !found {
+			extraA = append(extraA, element)
+		}
 	}
 
-	marshaller := &protojson.MarshalOptions{
-		Indent: "  ",
+	for j := 0; j < bLen; j++ {
+		if visited[j] {
+			continue
+		}
+		extraB = append(extraB, b[j])
 	}
 
-	s, err := marshaller.Marshal(m)
-	if err != nil {
-		return "", err
+	return
+}
+
+// formatSliceDiff is based on [assert.formatListDiff].
+func formatSliceDiff[V any, T message[V]](sliceA, sliceB, extraA, extraB []T) string {
```

Comment:
This formatting looks great! Still, we should see if we can exclude unexported fields.

Reply:
Looked into it, but I unfortunately could not find anything :(